### PR TITLE
feat: custom positioning for traffic light buttons

### PR DESCRIPTION
**Pull request by TonyWuu**
#### Description of Change
[There seems to be a high demand for custom inset traffic light buttons in Electron](https://github.com/electron/electron/issues/8237). 

- I added a new option in `BrowserWindow`, `trafficLightPosition` which sets the `(x,y)` position of the traffic light buttons. 
- It needs to be used in conjunction with `titleBarStyle: 'hidden'` to work. 
- Default to a value of (0, 0), which results in the standard default position, identical to the regular placement in `titleBarStyle: 'hidden'`. 

e.g.
```
titleBarStyle: 'hidden'
trafficLightPosition: {x: 12, y: 42}
```
image

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: add `trafficLightPosition` option in `BrowserWindow` API to allow custom positioning of traffic lights

### CODE REVIEW

File: shell/browser/native_window_mac.mm [hunk 1]
```diff
@@ -509,6 +511,46 @@ void ViewDidMoveToSuperview(NSView* self, SEL _cmd) {
     [NSEvent removeMonitor:wheel_event_monitor_];
 }
 
+void NativeWindowMac::RepositionTrafficLights() {
+  if (!traffic_light_offsetX_ && !traffic_light_offsetY_) {
+    NSLog(@"in here");
```
Comment (zcbenz) [inline shell/browser/native_window_mac.mm]: This can be removed.

File: shell/browser/native_window_mac.mm [hunk 2]
```diff
@@ -622,12 +664,23 @@ void ViewDidMoveToSuperview(NSView* self, SEL _cmd) {
 bool NativeWindowMac::IsVisible() {
   bool occluded = [window_ occlusionState] == NSWindowOcclusionStateVisible;
 
+  if (title_bar_style_ == TitleBarStyle::HIDDEN && !entering_fullscreen()) {
+    RepositionTrafficLights();
```
Comment (zcbenz) [inline shell/browser/native_window_mac.mm]: It does not seem appropriate to put this in `IsVisible`.

File: shell/browser/native_window_mac.mm [hunk 3]
```diff
@@ -509,6 +511,45 @@ void ViewDidMoveToSuperview(NSView* self, SEL _cmd) {
     [NSEvent removeMonitor:wheel_event_monitor_];
 }
 
+void NativeWindowMac::RepositionTrafficLights() {
+  if (!traffic_light_offsetX_ && !traffic_light_offsetY_) {
```
Comment (MarshallOfSound) [inline shell/browser/native_window_mac.mm]: This does not allow offsets set to `0` which someone may want

Comment (TonyWuu) [reply shell/browser/native_window_mac.mm]: My thinking was that an offset of 0 should result in the default position of the traffic lights (in `titleBarStyle: 'hidden'` mode). Are you suggesting there's a use case where someone might want an offset smaller than that default position? In that case, negative numbers would work here. 

IMO (x: 0, y: 0) should have a frame of reference relative to the default position, and not the absolute (0, 0) position of the frame

**Comment by jameskerr** [comment]
I'm excited for this! Thanks!

**Pull request merged**